### DISCUSSION TITLE
React on MCP23017 Buttons in Berry

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22416] [Q&A]
Hi.
When i used the PCF8574 for buttons, i do parse the message to check which button is/was pressed.
for example, i add one rule per PCF8574 to react on changes:
`tasmota.add_rule("PCF8574-2_INP", parse_trigger_msg)` 
and parse the Button with `def  parse_trigger_msg(value, trigger, msg) .... end`.

Now, playing with MCP23017, actually without interrupt pin connected. So I got the button informations dedicated for each button.
Like: ` 12:52:05.200 RSL: RESULT = {"Button9":{"Action":"SINGLE"}}` 

I could add a rule for each button number to reacton to the specific button, with "button<n>" as EventTrigger.
Since i plan to use 2 MCP23017 modules at least, i could result in 32 rules just for the buttons. What sounds to much rules, or less smart.
Is there an option to use the same trigger for each button, Something like "button?" instead of spexcifics "button1",button2",etc as triggers.

I saw with just a "?" as trigger, i could do a universal parse function, but this sounds not like the right way.

Will using the Interrupt of MCP23017 change the name of pinchange event  from "Button1/2/3.." to "McpInterruptA"?
So i could use one Trigger/Rule for 8 Inputs?

**Comment by sfromis** [comment]
Using Interrupt mode is about how input is processed, not to change how you use mode 2.

With Berry, it is no problem creating a wildcard rule, and do your own filtering by the features available in the language, including regular expressions. Just make sure to be specific about selecting what to process.

**Reply by ChrisVonFrosta** [reply]
Sorry, I don’t quite understand. Where did I ask for a change in Mode 2? I was asking about a different way of processing the input. At least that was my intention. 
I use mode 2 only to be able to use multiple MCP modules.

Unlike the PCF8574, the trigger event for the MCP23017 is individual for each button. If using the MCP interrupt, is a different status output, such as `RESULT = {"Button9":{"Action":"SINGLE"}}`? For example, would it be `RESULT = {"McpIntA":{"Button1":"SINGLE"}}`? Or should I understand that this output remains the same?

Could you give me a little nudge about the Berry wildcard? How could I respond to MCP inputs in Berry without having to create a rule for each button? Or actually with the "?" as rule trigger?

My shown rule example for the PCF8574 is the only way I know of to respond to PCF8574 inputs from Berry. This allows me to use one rule for all inputs. Parsing the input number is thus outside the rule, in a Berry function. I would like something similar for the MCP23017.

**Reply by sfromis** [reply]
You wrote:
> Will using the Interrupt of MCP23017 change the name of pinchange event

That would very much be a change of mode 2 behavior, to introduce different events. And mode 2 is indeed recommended as a good enhancement allowing the extra pins to work pretty much like standard gpios.

Since you posted about `?`, which indeed is the available wildcard, I assumed that you were already familiar with this, instead of one specific `tasmota.add_rule` for each. It is no problem grabbing all events this way, as long as you are careful about not having processing screwing up things. Like when it would be bad to have it cause similar events, which again would picked up ... ad infinitum.

And don't use `tasmota.cmd` (or `tasmota.set_power`) inside of the function on `tasmota.add_rule`, for such needs you should defer it till after returning, typically with
`tasmota.set_timer(0, /->tasmota.cmd ...)`

Beware that the existing non-Berry default actions will still apply even if you have a Berry `tasmota.add_rule` for the trigger, to avoid this, you may still need to decouple outside of Berry.

**Reply by ChrisVonFrosta** [reply]
> You wrote:
> 
> > Will using the Interrupt of MCP23017 change the name of pinchange event

Yes, but it was a question. It shoould mean: Will this happend if i use Interrupts?

But i got the answer out of here:
> .... And mode 2 is indeed recommended as a good enhancement allowing the extra pins to work pretty much like standard gpios.

> Since you posted about `?`, which indeed is the available wildcard, I assumed that you were already familiar with this, instead of one specific `tasmota.add_rule` for each. It is no problem grabbing all events this way, as long as you are careful about not having processing screwing up things. Like when it would be bad to have it cause similar events, which again would picked up ... ad infinitum.
Ok, you mean the same mechanism with ?-rule. I could imagine it. Wasn't sure, if you know better options.

> And don't use `tasmota.cmd` (or `tasmota.set_power`) inside of the function on `tasmota.add_rule`, for such needs you should defer it till after returning, typically with `tasmota.set_timer(0, /->tasmota.cmd ...)`
thx! Anyway, i would define functions to be triggered by rule. And keep enclosures in mind.
 
> Beware that the existing non-Berry default actions will still apply even if you have a Berry `tasmota.add_rule` for the trigger, to avoid this, you may still need to decouple outside of Berry.
good to keep in mind.


Would you go with a wildcard rule or a seprate `tasmota.add_rule()` for each button number? I would prefere the wildcard. but than the parsing function would be procceed frquently even not button was pressed.

**Reply by sfromis** [reply]
Yes, you asked a question about you needing to use mode 2 in a very different way with interrupts. And I answered no to that, it does not emit different events.

If needing to handle "many" Button events for MCP23017 mode 2, I'd go for a wildcard rule, and just be a bit careful about coding the handler to handle "everything", while being very specific about filtering what events it wants to process. You should not worry too much about executing a few lines of simple code in Berry, that's not something bogging an ESP32 down. And it is not really a **lot** of events during normal operation anyway.

**Reply by ChrisVonFrosta** [reply]
thank you for giving the green light (confirmation) for parsing the wildcard ;)

**Reply by ChrisVonFrosta** [reply]
> including regular expressions.

I can't find anz information about using regular expressions in in berry.
But i did it with `string.find` and `string.split`

**Reply by sfromis** [reply]
Here's the docs for regular expressions in Berry: https://tasmota.github.io/docs/Berry/#re-regex-module
Anyway, neither regex, find nor split are "needed" for handling the wildcards, see the example I posted recently here.

**Reply by ChrisVonFrosta** [reply]
but you use one find in your example?! ;) anyway thx!!!

**Reply by sfromis** [reply]
I do not use `string.find` (which was what you mentioned) in my example, only `map.find`. The point was that the string processing part is pretty simple for a prefix.

**Reply by ChrisVonFrosta** [reply]
understood. but why not using import string-things? performance? I use STRING in the following script either way.

**Reply by sfromis** [reply]
> using import string-things

What does those words even mean? 

My point was that using `string.find` and `string.split` for the wildcard handling would likely be less simple than my example.

When it comes to `import`, that is not something you should worry about performance-wise, as it is to tell the compiler where to find stuff. My preference is to do it in the individual functions, to avoid having functions depends on things done at the global level, hence I have the `import string` in the function. My style is to avoid doing things in a wider scope than required, but I'm not saying "should".

**Comment by barbudor** [comment]
> Unlike the PCF8574, the trigger event for the MCP23017

PCF8574 also has a mode 2 which allows to use it as buttons/switches/relays

**Reply by ChrisVonFrosta** [reply]
> > Unlike the PCF8574, the trigger event for the MCP23017
> 
> PCF8574 also has a mode 2 which allows to use it as buttons/switches/relays

Oh that is new for me.
Maybe I missunderstand the Mode2. I came to it just to use more than one MCP23017 module. I thought that is the main reason of Mode2, because I used Mode2 with the PCA9685 (PWM controller) for this reason.

So, you say, that mode2 in relation to MCP23017 and PCF8574, is mainly for changing the behavior to nativ buttons and relays?
So I could also use more than one module in mode1?

**Reply by sfromis** [reply]
Mode 2 is indeed mainly to be a **significant** enhancement over how it works with mode 1, as you simply can treat expander pins as a normal button/switch/relay/output. With mode 2 available, I don't see much reason to even use mode 1.

**Reply by ChrisVonFrosta** [reply]
> PCF8574 also has a mode 2 which allows to use it as buttons/switches/relays

You know, I am amazed anew each time by what you guys achieve here. And I know that with Berry I practically have all possibilities. But if you would now allow Tasmota to handle not just multiple relays, buttons, etc., but also several (CCT) lamps, it would probably greatly shorten my Berry code and also perform better than my Berry solution. Even though I'm not the only one who has expressed such a wish, I understand that it might still not be relevant.

You had already explained the philosophy behind it. But I still don't want to accept it ;) Thanks again!

**Reply by sfromis** [reply]
The "issue" with multiple lights (which it would be for multiple CCT) is not "philosophy", but no one seems interested in opening that can of worms. The lights driver got pretty complex and reorganizing that "jenga tower" sounds like a recipe for pain.....

**Reply by ChrisVonFrosta** [reply]
on which files dependsthe light/Output stuctur/driver? Just to take a look and run away :)

**Reply by sfromis** [reply]
AFAIK, mainly https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_04_light.ino

**Run!** :fearful:

**Comment by ChrisVonFrosta** [comment]
hi, i have to ask again, sorry.

I would go with wildcar "?" but i dislike to grab the number of input out of msg string for example "Button9".  
Input msg:
`19:47:58.868 RSL: RESULT = {"Button9":{"Action":"SINGLE"}}`

It would be more easier for me to have the number in its own datafield.
like:
`19:47:58.868 RSL: RESULT = {"Button":{"no":9,"Action":"SINGLE"}}`
or 
`19:47:58.868 RSL: RESULT = {"Input":{"Button":9,"Action":"SINGLE"}}`

Where is this output text defined? I took a look to `xdrv_67_mcp23xxx.ino` but i suspected that it is generated elsewhere.
Would it be a hard break, to the button system if i edit just this Outbut text?

**Reply by sfromis** [reply]
Well, it's not *that* hard to use the wildcard method. Test example:
```
tasmota.remove_rule('?')
tasmota.add_rule('?',
  def (value,trigger,data)
    if type(data) != 'instance' return end
    var key = data.keys()()
    import string
    if ! string.startswith(key,'Button') return end
    var button = int(key[6..])
    var state = data[key].find('State',-1)
    print(button, state)
  end)
```

**Reply by ChrisVonFrosta** [reply]
```
tasmota.remove_rule('?')
def parse_trigger_SYSTEMBUTTON2(value,trigger,data)
    if type(data) != 'instance' return end
    var key = data.keys()()
    import string
    if ! string.startswith(key,'Button') return end
    var button = int(key[6..])
    var state = data[key].find('State',-1)
    print(button, state)
end
tasmota.add_rule("?", parse_trigger_SYSTEMBUTTON2)
	:5: in function `parse_trigger_SYSTEMBUTTON2`
	<unknown source>: in function `try_rule`
	<unknown source>: in function `exec_rules`
	<unknown source>: in function `event`
BRY: Exception> 'attribute_error' - module 'string' has no member 'startswith'
```

**Reply by ChrisVonFrosta** [reply]
```
string.split
<function: 0x400fff58>
string.startswith
BRY: Exception> 'attribute_error' - module 'string' has no member 'startswith'
stack traceback:
	:1: in function `main`
```

**Reply by sfromis** [reply]
I was assuming that you use current Tasmota, instead of lingering on an outdated version without recently introduced features.

`string.startswith(key,'Button')` can also be written as `string.find(key,'Button') == 0`

**Comment by ChrisVonFrosta** [comment]
fu**. I did not noticed that i was working on the wrong tab(with other old remote device for same purpose). damn. i was just going crazy why the pininput didn't recogniezed anymore. *so stupid*
I am so sorry. ignore my last posts. i will check aggain
